{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Standard primers for the Yeast Pathway Kit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The six primers below are necessary and sufficient to carry out assemblies with the YPK system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "from pydna.parsers import parse\n",
    "p567,p577,p468,p467,p568,p578,p775,p778  =       parse(''' >567_pCAPsAjiIF (23-mer)\n",
    "                                                            GTcggctgcaggtcactagtgag\n",
    "                                                            >577_crp585-557 (29-mer)\n",
    "                                                            gttctgatcctcgagcatcttaagaattc\n",
    "\n",
    "                                                            >468_pCAPs_release_fw (25-mer)\n",
    "                                                            gtcgaggaacgccaggttgcccact\n",
    "                                                            >467_pCAPs_release_re (31-mer) \n",
    "                                                            ATTTAAatcctgatgcgtttgtctgcacaga\n",
    "\n",
    "                                                            >568_pCAPsAjiIR (22-mer) \n",
    "                                                            GTGCcatctgtgcagacaaacg\n",
    "                                                            >578_crp42-70 (29-mer)\n",
    "                                                            gttcttgtctcattgccacattcataagt\n",
    "\n",
    "                                                            >775_tp_ZraI_fwd (18-mer)\n",
    "                                                            gcggccgctgacTTAAAT\n",
    "                                                            >778_tp_Eco32I_rev (20-mer)\n",
    "                                                            ggtaaatccggatTAATTAA''')\n",
    "\n",
    "# The primers 167 and 166 were used instead of 577 and 578 for some constructions\n",
    "\n",
    "p167,p166 =       parse(''' >167 pCAPSfw (24-mer)\n",
    "                            TCCTGACGGGTAATTTTGATTTGC\n",
    "                            >166 pCAPSrv (24-mer) \n",
    "                            CTGTGAAGTGGCTGAAATTTCGTA''')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Primer positions in pYPKa derived vectors\n",
    "\n",
    "                                 >-gene-->\n",
    "                >-TP-->           \\     /           >-TP-->\n",
    "                 \\   /             \\   /             \\   /\n",
    "                  \\ /               \\ /               \\ /\n",
    "         167>  511>|<776             |             777>|<512 <166\n",
    "     577>      775>|468>         <567|568>         <467|<778      <578\n",
    "                   |                 |                 | \n",
    "     ✽✽✽✽✽✽✽✽✽✽✽✽N-Z=================A+++++++++++++++++E-A••••••••••••••••••••\n",
    "    |            o r                 j                 c c                    |\n",
    "    |            t a                 i                 o c                    |\n",
    "    |            I I                 I                 R I                    |\n",
    "    |                               (*)                V I                    |\n",
    "    |                                                    I                    |\n",
    "    |                                                                         |\n",
    "     --------------------- pYPKa ---------------------------------------------\n",
    "    \n",
    "    (* )AjiI is only unique in the pYPKa vector, not in the pYPK0 or vectors derived from pYPK0 such as pYPKpw   vectors.\n",
    "    \n",
    "    NotI   GC^GGCCGC\n",
    "    ZraI   GAC^GTC\n",
    "    AjiI   CAC^GTC\n",
    "    EcoRV  GAT^ATC\n",
    "    AccIII T^CCGGA\n",
    "    \n",
    "    \n",
    "    Below is a list of primers for the pYPKa backbone that have been tested\n",
    "    \n",
    "    577 gttctgatcctcgagcatcttaagaattc\n",
    "    167 TCCTGACGGGTAATTTTGATTTGC\n",
    "    511 gtcagcggccgcattgcacagactctgct\n",
    "    775 gcggccgctgacTTAAAT\n",
    "    776 cgttcctcgacTAATTAA\n",
    "    468 gtcgaggaacgccaggttgcccact\n",
    "    567 GTcggctgcaggtcactagtgag\n",
    "    568 GTGCcatctgtgcagacaaacg\n",
    "    467 ATTTAAatcctgatgcgtttgtctgcacaga\n",
    "    777 acgcatcaggatTTAAAT\n",
    "    778 ggtaaatccggatTAATTAA\n",
    "    512 ttcgccaattgattcaggtaaatccggat\n",
    "    578 gttcttgtctcattgccacattcataagt\n",
    "    166 CTGTGAAGTGGCTGAAATTTCGTA\n",
    "    578 gttcttgtctcattgccacattcataagt\n",
    "    342 CCTTTTTACGGTTCCTGGCCT\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "### Background\n",
    "\n",
    "The yeast pathway kit is a system for the _in vivo_ assembly of entire metabolic or other multi gene pathways in _Saccharomyces cerevisiae_. It has three principal components which are combinations of DNA fragments cloned in the pYPKa positive selection vector:\n",
    "\n",
    "1. pYPKa clones of promoters cloned using the blunt ZraI restriction site\n",
    "2. pYPKa clones of genes to be expressed cloned using the blunt AjiI restriction site\n",
    "3. pYPKa clones of terminators cloned using the blunt EcoRV restriction site\n",
    "4. pYPKpw acceptor vector\n",
    "\n",
    "### pYPKa\n",
    "\n",
    "pCAPs [Schlieper et al., 1998] is a small (3128 bp) positive selection vector for E. coli. It has the the usual ampicilllin resistance gene for selection and a collection of restriction sites in a gene encoding a toxic version of the cyclic AMP receptor protein, CRP. When DNA fragments are cloned within this gene, cell survival is permitted by the inactivation of the toxic CRP gene. Thus, there is positive selection for inserted fragments within the gene. This leads to very high frequency of correct clones in cloning experiments. pYPKa is a modified version carrying an additional AjiI restriction site.\n",
    "\n",
    "The pYPKa CRP gene has the following useful restriction enzyme sites in the order they appear in the pCAPs sequence file which is the opposite direction of the transcription of the CRP gene. The positions are given as distance between cut cites of the enzymes.\n",
    "\n",
    "\n",
    "| Enzyme | distance |\n",
    "|--------|-----|\n",
    "| stop   | 0   |\n",
    "| XhoI   | 58  |\n",
    "| EcoRI  | 14  |\n",
    "| XbaI   | 19  |\n",
    "| MscI   | 21  |\n",
    "| SphI   | 33  |\n",
    "| NotI   | 60  |\n",
    "| **ZraI**   | 11  |\n",
    "| **AjiI**   | 50  |\n",
    "| **EcoRV**  | 32  |\n",
    "| AccIII |  2  |\n",
    "| MfeI   | 17  |\n",
    "| BsaAI  | 74  |\n",
    "| OliI   | 164 |\n",
    "| BamHI  | 70  |\n",
    "| start  | 28  |\n",
    "\n",
    "\n",
    "## Strategy for cloning promoters and terminators in pYPKa\n",
    "\n",
    "The same genetic elements are used as both promoters and terminators in the yeast pathway kit system. This is possible if intergenic sequences from between tandemly expressed genes are used.\n",
    "\n",
    "Tandemly expressed genes:\n",
    "\n",
    "    ... Promoter> Gene 1 >Terminator Promoter> Gene 2 >Terminator ...\n",
    "\n",
    "The first part of such an intergenic sequence contains the terminator of the upstream gene, and the latter part the promoter of the downstream gene.\n",
    "\n",
    "There are at least seven promoters that include strong promoters that have been reportedly used for expression of homologous or heterologous genes in S. cerevisiae (see table below) that are also intergenic sequences from tandemly expressed genes. These range in size between 500 and 900 bp.\n",
    "\n",
    "| TP     | Expressed gene                             |\n",
    "|--------|--------------------------------------------|\n",
    "| ScENO2 | Enolase II                                 |\n",
    "| ScFBA1 | Fructose 1,6-bisphosphate aldolase         |\n",
    "| ScPDC1 | Pyruvate decarboxylase                     |\n",
    "| ScPGI1 | Phosphoglucose isomerase                   |\n",
    "| ScTDH3 | Glyceraldehyde-3-phosphate-dehydrogenase   |\n",
    "| ScTEF1 | Translational elongation factor EF-1 alpha |\n",
    "| ScTPI1 | Triose phosphate isomerase                 |\n",
    "\n",
    "Each promoter is amplified with primers with six base pairs overhang. The overhangs\n",
    "are identical for all promoters and terminators.\n",
    "\n",
    "\n",
    "Figure 2: Tails of primers used to amplify tps\n",
    "\n",
    "    >ScNNNNtpf\n",
    "    TTAAAT.....\n",
    "\n",
    "    >ScNNNNtpr_PacI\n",
    "    TAATTAA.....\n",
    "\n",
    "Each PCR product will have the following general structure (both strands shown):\n",
    "\n",
    "        TTAAAT...TTAATTA\n",
    "        AATTTA...AATTAAT\n",
    "\n",
    "\n",
    "The forward (tpf) primer adds a partial SmiI (ATTT^AAAT) site in front of the terminator of the tp PCR product. The reverse (tpr_PacI) primer adds a partial PacI (TTAAT^TAA) site at the end of the promoter. These sites are only present when cloned in the EcoRV site in pYPKa location as shown below:\n",
    "\n",
    "        ZraI                                              AjiI                           EcoRV\n",
    "\n",
    "    ===GAC                                                                                   \n",
    "          GTCgaggaacgccaggttgcccactttctcactagtgacctgcagccGAC                               \n",
    "                                                            GTGccatctgtgcagacaaacgcatcagGAT\n",
    "                                                                                           ATCcggattt---\n",
    "\n",
    "pYPKa opened by ZraI or EcoRV:\n",
    "\n",
    "                ZraI                                EcoRV \n",
    "\n",
    "    ===GAC                   GTC===AjiI===GAT                   ATC---\n",
    "    \n",
    "    \n",
    "Terminator promoter sequences cloned in both locations:\n",
    "    \n",
    "\n",
    "    ===GAC TTAAAT-promoter-TTAATTA GTC===AjiI===GAT TTAAAT-terminator-TTAATTA ATC---\n",
    "                                                 ---------            ---------\n",
    "                                           \n",
    "                                                    SmiI                 PacI\n",
    "\n",
    "The [SmiI](http://rebase.neb.com/rebase/enz/SmiI.html) site can be used to add for example c-terminal tags or as an alternative for pathway extension.\n",
    "\n",
    "## Gene cloning\n",
    "\n",
    "\n",
    "Tailed primers can be used to amplify genes for recombination between any two promoters and terminators. The primer tails need to be designed to promote recombination between the promoter and terminator:\n",
    "\n",
    "     >Gene_YPK_rec_fwd\n",
    "     tgcccactttctcactagtgacctgcagccgac-A-ATG...\n",
    "                                         ---\n",
    "                                         sta\n",
    "                                         rtc\n",
    "                                         odo\n",
    "                                         n                                              \n",
    "     <--------------  33 nt -------->\n",
    "\n",
    "     >Gene_YPK_rec_rev\n",
    "     AAatcctgatgcgtttgtctgcacagatggCAC...\n",
    "                                      ---\n",
    "                                      sto\n",
    "                                      pco\n",
    "                                      don\n",
    "     <---------- 33 nt--------------->\n",
    "\n",
    "Alternatively, genes can be cloned in the blunt AjiI site of pYPKa by the same strategy as used for the promoters and terminators.\n",
    "\n",
    "The gene can be released by ZraI + EcoRV digestion if no restriction sites for these genes are present in the gene, alternatively the gene with flanking sequences can be reamplified with primers 468 and 467.\n",
    "\n",
    "The resulting cassette is similar to the one amplified with the tailed primers above, but the cloning primers can be made shorter and thereby less expensive and possibly more efficient.\n",
    "\n",
    "The primers 775, 776, 777 and 778 can be used to check presence and orientation of all\n",
    "terminator-promoters cloned in the ZraI or EcoRV site in pCAPs\n",
    "\n",
    "\n",
    "## pYPKpw\n",
    "\n",
    "The pYPKpw is a version of the pCAPs vector that has additional sequences for maintenance in _Saccharomyces cerevisiae_ . The additional genes are URA3 for prototrophic selection for uracile and the 2µ sequence for replication There is also a deletion in the CRP gene.\n",
    "\n",
    "\n",
    "\n",
    "## Assembly of single gene expression vectors\n",
    "\n",
    "The pYPKpw plasmid is linearized with ZraI, FspAI or EcoRV and mixed with three linear DNA fragments:\n",
    "\n",
    "* A promoter amplified with the 577 and 567 primer\n",
    "* a gene amplified with tailed primers or using 467, 468\n",
    "* A terminator amplified with the 568 and 578 primers\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "           ✽✽✽✽promoter====    ++++terminator••••\n",
    "            \\/          \\/      \\/            \\/\n",
    "            /\\          /\\      /\\            /\\\n",
    "     ------✽✽✽✽        ====gene++++          ••••-------                           \n",
    "    |                                                   |\n",
    "    |                                                   |\n",
    "    |                                                   |\n",
    "     --------------------- pYPKpw ----------------------\n",
    "\n",
    "\n",
    "The entire single gene expression cassette (promoter-gene-terminator) can be cut out from the pYPKpw vector using [NotI](http://rebase.neb.com/rebase/enz/NotI.html) on the plasmid backbone and the [PacI](http://rebase.neb.com/rebase/enz/PacI.html) generated by the tp primer tails."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
